packages/authorization-agent/src/authorization-agent.ts

```diff
@@ -130,4 +183,109 @@ export class AuthorizationAgent {
       affectedAuthorizations.map(async (accessAuthorization) => this.generateAccessGrant(accessAuthorization.iri))
     );
   }
+
+  public async findSocialAgentsWithAccess(dataInstanceIri: string): Promise<AgentWithAccess[]> {
+    const dataInstance = await this.factory.readable.dataInstance(dataInstanceIri);
+    const shapeTree = dataInstance.dataRegistration.registeredShapeTree;
+    const agentsWithAccess: AgentWithAccess[] = [];
+    for await (const accessAuthorization of this.accessAuthorizations) {
+      const dataAuthorization = (
+        await iterable2array<ReadableDataAuthorization>(accessAuthorization.dataAuthorizations)
+      ).find((autorization) => autorization.registeredShapeTree === shapeTree);
+      // eslint-disable-next-line no-continue
+      if (!dataAuthorization) continue;
+
+      switch (dataAuthorization.scopeOfAuthorization) {
+        case INTEROP.All.value:
+          agentsWithAccess.push(formatAgentWithAccess(dataAuthorization));
+          break;
+        case INTEROP.AllFromAgent.value:
+          // TODO: rethink for delegated sharing, e.g. Alice shares project owned by ACME
+          if (dataAuthorization.dataOwner === this.webId) {
+            agentsWithAccess.push(formatAgentWithAccess(dataAuthorization));
+          }
+          break;
+        case INTEROP.AllFromRegistry.value:
+          if (dataAuthorization.hasDataRegistration === dataInstance.dataRegistration.iri) {
+            agentsWithAccess.push(formatAgentWithAccess(dataAuthorization));
+          }
+          break;
+        case INTEROP.SelectedInstances.value:
+          if (
+            dataAuthorization.hasDataRegistration === dataInstance.dataRegistration.iri &&
+            dataAuthorization.hasDataInstance.includes(dataInstanceIri)
+          ) {
+            agentsWithAccess.push(formatAgentWithAccess(dataAuthorization));
+          }
+          break;
+        default:
+          throw new Error(
+            `encountered incorect Data Authorization with scope:${dataAuthorization.scopeOfAuthorization}`
+          );
+      }
+    }
+    const socialAgentsWithAccess: AgentWithAccess[] = [];
+    for await (const registration of this.socialAgentRegistrations) {
+      const socialAgentWithAccess = agentsWithAccess.find(({ agent }) => agent === registration.registeredAgent);
+      if (socialAgentWithAccess) {
+        socialAgentsWithAccess.push(socialAgentWithAccess);
+      }
+    }
+    return socialAgentsWithAccess;
+  }
+
+  /*
+   * Authorizes access to a specific Data Instance to multiple agents
+   * TODO: support delegated authorization
+   */
+  public async shareDataInstance(details: ShareDataInstanceStructure): Promise<void> {
```

Comment:
Also, while I like compressing parameters into option objects in this case it obfuscates the inputs needed by the function. I'd rather have as many parameters as needed here, specially because the type name `ShareDataInstanceStructure` does not convey at all what it actually is.

Reply:
I'm going to create an issue to refactor all the public interfaces, this one pretty much duplicates one in sai-api-messages since the data needs to come from user interaction. Given all the functionality in place and experience with implementing both authorizations agent and demo applications, we can make quite a lot of improvement.
I'm going to link that issue to your comment as one of the relevant suggestions.

Reply:
Tracked in #54